File: calculate_overlap.m
function [equalization_overlap] = calculate_overlap(cell_values)

    % global no_ovp_plot_flag;
	%if isempty(no_ovp_plot_flag)
	%	no_ovp_plot_flag = 0;
	%end

    no_ovp_plot_flag = 0;

    % Find local maxima (peaks) and minima (valleys)
    is_peak = islocalmax(cell_values);
    is_valley = islocalmin(cell_values);
    
    peak_indices = find(is_peak);
    valley_indices = find(is_valley);

    % Calculate local heights of peaks
    local_heights = zeros(size(peak_indices));
    for i = 1:length(peak_indices)
        left_valleys = valley_indices(valley_indices < peak_indices(i));
        right_valleys = valley_indices(valley_indices > peak_indices(i));
        if isempty(left_valleys)
            left_min = cell_values(1);
        else
            left_min = cell_values(left_valleys(end));
        end
        if isempty(right_valleys)
            right_min = cell_values(end);
        else
            right_min = cell_values(right_valleys(1));
        end
        local_heights(i) = cell_values(peak_indices(i)) - max(left_min, right_min);
    end

    % Calculate local depths of valleys
    local_depths = zeros(size(valley_indices));
    for i = 1:length(valley_indices)
        left_peaks = peak_indices(peak_indices < valley_indices(i));
        right_peaks = peak_indices(peak_indices > valley_indices(i));
        if isempty(left_peaks)
            left_max = cell_values(1);
        else
            left_max = cell_values(left_peaks(end));
        end
        if isempty(right_peaks)
            right_max = cell_values(end);
        else
            right_max = cell_values(right_peaks(1));
        end
        local_depths(i) = min(left_max, right_max) - cell_values(valley_indices(i));
    end

    % Calculate equalization overlap
    equalization_overlap = sum(local_heights) + sum(local_depths);
    
    % Plot cell_values with peaks and valleys highlighted
    % plot(cell_values)
    % hold on
    % plot(peak_indices, cell_values(peak_indices), 'ro')
    % plot(valley_indices, cell_values(valley_indices), 'bo')
    % hold off

end

File: codegen/lib/pso/calculate_overlap.c
#include <stdio.h>
#include <stdbool.h>
#include <math.h>

// Define maximum size for arrays
#define MAX_SIZE 100

// Helper function to find local maxima
void find_local_max(const double data[], int size, bool is_peak[]) {
    for (int i = 1; i < size - 1; ++i) {
        if (data[i] > data[i - 1] && data[i] > data[i + 1]) {
            is_peak[i] = true;
        } else {
            is_peak[i] = false;
        }
    }
    is_peak[0] = false;
    is_peak[size - 1] = false;
}

// Helper function to find local minima
void find_local_min(const double data[], int size, bool is_valley[]) {
    for (int i = 1; i < size - 1; ++i) {
        if (data[i] < data[i - 1] && data[i] < data[i + 1]) {
            is_valley[i] = true;
        } else {
            is_valley[i] = false;
        }
    }
    is_valley[0] = false;
    is_valley[size - 1] = false;
}

// Function to calculate the valid size of the input array
int calculate_valid_size(const double data[], int max_size) {
    int valid_size = 0;
    for (int i = 0; i < max_size; ++i) {
        if (data[i] >= 0.1 && data[i] <= 100.1) {
            valid_size++;
        } else {
            break;
        }
    }
    return valid_size;
}

// Function to calculate overlap
double calculate_overlap(const double cell_values[], int max_size[]) {
    int size = calculate_valid_size(cell_values, *(max_size + 1));
    bool is_peak[MAX_SIZE] = {false};
    bool is_valley[MAX_SIZE] = {false};
    int peak_indices[MAX_SIZE] = {0};
    int valley_indices[MAX_SIZE] = {0};
    double local_heights[MAX_SIZE] = {0.0};
    double local_depths[MAX_SIZE] = {0.0};
    int peak_count = 0;
    int valley_count = 0;

    find_local_max(cell_values, size, is_peak);
    find_local_min(cell_values, size, is_valley);

    for (int i = 0; i < size; ++i) {
        if (is_peak[i]) {
            peak_indices[peak_count++] = i;
        }
        if (is_valley[i]) {
            valley_indices[valley_count++] = i;
        }
    }

    for (int i = 0; i < peak_count; ++i) {
        int peak_idx = peak_indices[i];
        double left_min = cell_values[0];
        double right_min = cell_values[size - 1];

        for (int j = 0; j < valley_count; ++j) {
            if (valley_indices[j] < peak_idx) {
                left_min = fmin(left_min, cell_values[valley_indices[j]]);
            } else if (valley_indices[j] > peak_idx) {
                right_min = fmin(right_min, cell_values[valley_indices[j]]);
                break;
            }
        }
        local_heights[i] = cell_values[peak_idx] - fmax(left_min, right_min);
    }

    for (int i = 0; i < valley_count; ++i) {
        int valley_idx = valley_indices[i];
        double left_max = cell_values[0];
        double right_max = cell_values[size - 1];

        for (int j = 0; j < peak_count; ++j) {
            if (peak_indices[j] < valley_idx) {
                left_max = fmax(left_max, cell_values[peak_indices[j]]);
            } else if (peak_indices[j] > valley_idx) {
                right_max = fmax(right_max, cell_values[peak_indices[j]]);
                break;
            }
        }
        local_depths[i] = fmin(left_max, right_max) - cell_values[valley_idx];
    }

    double equalization_overlap = 0.0;
    for (int i = 0; i < peak_count; ++i) {
        equalization_overlap += local_heights[i];
    }
    for (int i = 0; i < valley_count; ++i) {
        equalization_overlap += local_depths[i];
    }

    return equalization_overlap;
}

